### PR TITLE
Search query parsing: fix bug in `reduce` causing issues with search contexts

### PR DESCRIPTION
**Pull request by camdencheek**
Fixes https://github.com/sourcegraph/sourcegraph/issues/35096

Explanation of changes inline. 

First commit adds a minimal test case reproducing the issue. Second commit adds a fix with updated test output.

## Test plan

Added a test. We may want to consider testing this a bit more rigorously, but this will cover regression.

### CODE REVIEW

File: internal/search/query/parser.go [hunk 1]
```diff
@@ -1021,9 +1021,11 @@ func reduce(left, right []Node, kind operatorKind) ([]Node, bool) {
...
 			return left, true
 		}
-		if operator, ok := left[0].(Operator); ok && operator.Kind == kind {
-			// Reduce left node.
-			return append(operator.Operands, right...), true
+		if len(left) == 1 {
+			if operator, ok := left[0].(Operator); ok && operator.Kind == kind {
+				// Reduce left node.
+				return append(operator.Operands, right...), true
+			}
```
Comment (camdencheek) [inline internal/search/query/parser.go:1028]: The issue here is that line `1025` assumes that `left` is only a single node. This is true whenever we call this function at the top level, but the recursive call on line `1047` will grow `left` to be multiple nodes. When `len(left) > 0`, this block of logic will merge `right` into `left[0]`, then return the resulting set of operands, effectively dropping `left[1:]`.

**Review by rvantonder — COMMENTED** [review]
So the `len` check doesn't feel quite like the right bandaid. The way to think about reduce, and the style of the pass, is that the left side is always in the reduced state. There are two ways that preserve this style of the implementation (though the fix is still incomplete). One is to compare the last element in the left tree:

```patch
@@ -1021,11 +1021,9 @@ func reduce(left, right []Node, kind operatorKind) ([]Node, bool) {
 			}
 			return left, true
 		}
-		if len(left) == 1 {
-			if operator, ok := left[0].(Operator); ok && operator.Kind == kind {
-				// Reduce left node.
-				return append(operator.Operands, right...), true
-			}
+		if operator, ok := left[len(left)-1].(Operator); ok && operator.Kind == kind {
+			// Reduce left node.
+			return append(operator.Operands, right...), true
 		}
```

The other is to maintain the invariant that we call reduce on at most one left-most reduced node, which I think I prefer, so that we can do the complete fix:

```patch
@@ -1040,9 +1040,9 @@ func reduce(left, right []Node, kind operatorKind) ([]Node, bool) {
 	}
 	if len(right) > 1 {
 		// Reduce right list.
-		reduced, changed := reduce(append(left, right[0]), right[1:], kind)
+		reduced, changed := reduce([]Node{right[0]}, right[1:], kind)
 		if changed {
-			return reduced, true
+			return append(left, reduced...), true
 		}
 	}
 	return append(left, right...), false
```


I think the complete fix (to always reduce) needs to loop over the terms in the right list and accumulate/merge in this block based on operator kind. I can look at it later.

Agree/disagree? I'm OK with approving but prefer avoiding the `len` check.

**Comment by camdencheek** [comment]
Nice, agreed. The second option definitely feels better. Seems like maybe we should make that invariant (`len(left) == 1`) part of the type signature, but that is more effort than I want to put into this right now. 

I pushed a commit with that change

**Review by rvantonder — APPROVED** [review]
nice. I'll bisect the regression later so we can add changelog for affected versions

**Pull request merged**